modules/ml/src/svmsgd.cpp

```diff
@@ -0,0 +1,595 @@
+        marginTypeStr = format("Unknown_%d", params.marginType);
+    default:
+        std::cout << "params.marginType isn't initialized" << std::endl;
+    }
+
+    fs << "marginType" << marginTypeStr;
+
+    fs << "lambda" << params.lambda;
+    fs << "gamma0" << params.gamma0;
+    fs << "c" << params.c;
+
+    fs << "term_criteria" << "{:";
+    if( params.termCrit.type & TermCriteria::EPS )
+        fs << "epsilon" << params.termCrit.epsilon;
+    if( params.termCrit.type & TermCriteria::COUNT )
+        fs << "iterations" << params.termCrit.maxCount;
+    fs << "}";
+}
+
+void SVMSGDImpl::read(const FileNode& fn)
+{
+    clear();
+
+    readParams(fn);
+
+    fn["weights"] >> weights_;
+    fn["shift"] >> shift_;
+}
+
+void SVMSGDImpl::readParams( const FileNode& fn )
+{
+    String svmsgdTypeStr = (String)fn["svmsgdType"];
+    SvmsgdType svmsgdType =
+            svmsgdTypeStr == "SGD" ? SGD :
+                                     svmsgdTypeStr == "ASGD" ? ASGD : ILLEGAL_SVMSGD_TYPE;
+
+    if( svmsgdType == ILLEGAL_SVMSGD_TYPE )
+        CV_Error( CV_StsParseError, "Missing or invalid SVMSGD type" );
+
+    params.svmsgdType = svmsgdType;
+
+    String marginTypeStr = (String)fn["marginType"];
+    MarginType marginType =
+            marginTypeStr == "SOFT_MARGIN" ? SOFT_MARGIN :
+                                     marginTypeStr == "HARD_MARGIN" ? HARD_MARGIN : ILLEGAL_MARGIN_TYPE;
+
+    if( marginType == ILLEGAL_MARGIN_TYPE )
+        CV_Error( CV_StsParseError, "Missing or invalid margin type" );
+
+    params.marginType = marginType;
+
+    CV_Assert ( fn["lambda"].isReal() );
+    params.lambda = (float)fn["lambda"];
+
+    CV_Assert ( fn["gamma0"].isReal() );
+    params.gamma0 = (float)fn["gamma0"];
+
+    CV_Assert ( fn["c"].isReal() );
+    params.c = (float)fn["c"];
+
+    FileNode tcnode = fn["term_criteria"];
+    if( !tcnode.empty() )
+    {
+        params.termCrit.epsilon = (double)tcnode["epsilon"];
+        params.termCrit.maxCount = (int)tcnode["iterations"];
+        params.termCrit.type = (params.termCrit.epsilon > 0 ? TermCriteria::EPS : 0) +
+                (params.termCrit.maxCount > 0 ? TermCriteria::COUNT : 0);
+    }
+    else
+        params.termCrit = TermCriteria( TermCriteria::EPS + TermCriteria::COUNT, 100000, FLT_EPSILON );
+
+}
+
+void SVMSGDImpl::clear()
+{
+    weights_.release();
+    shift_ = 0;
+}
+
+
+SVMSGDImpl::SVMSGDImpl()
+{
+    clear();
+
+    params.svmsgdType = ILLEGAL_SVMSGD_TYPE;
+    params.marginType = ILLEGAL_MARGIN_TYPE;
+
+    // Parameters for learning
+    params.lambda = 0;                              // regularization
+    params.gamma0 = 0;                        // learning rate (ideally should be large at beginning and decay each iteration)
+    params.c = 0;
+
+    TermCriteria _termCrit(TermCriteria::COUNT + TermCriteria::EPS, 0, 0);
+    params.termCrit = _termCrit;
+}
+
+void SVMSGDImpl::setOptimalParameters(int svmsgdType, int marginType)
```

Comment:
I would think, that this function should be called in constructor. Is there a reason not to?